pkg/authorization/admission/restrictnodes/restrictnodes.go

```
@@ -0,0 +1,222 @@
+		return q.admitPod(nodeName, a)
+	default:
+		return nil
+	}
+}
+
+var (
+	everything = labels.Everything()
+)
+
+func (q *restrictNodesAdmission) admitSecret(nodeName string, a admission.Attributes) error {
+	ns := a.GetNamespace()
+	name := a.GetName()
+	pods, err := q.pods.Pods(ns).List(everything)
+	if err != nil {
+		return err
+	}
+	for _, pod := range pods {
+		if pod.Spec.NodeName != nodeName {
+			continue
+		}
+		found := false
+		visitSecretNames(pod, func(referencedName string) bool {
+			if name == referencedName {
+				found = true
+				return false
+			}
+			return true
+		})
+		if found {
+			return nil
+		}
+	}
+
+	// TODO: other secret references nodes need? PVs?
+
+	return kapierrors.NewForbidden(a.GetResource().GroupResource(), a.GetName(), errors.New("cannot access secret"))
+}
+
+func visitSecretNames(pod *kapi.Pod, visitor func(string) bool) {
+	for _, s := range pod.Spec.ImagePullSecrets {
+		if !visitor(s.Name) {
+			return
+		}
+	}
+	for _, v := range pod.Spec.Volumes {
+		if v.Secret != nil && !visitor(v.Secret.SecretName) {
+			return
+		}
+	}
+	for _, v := range pod.Spec.InitContainers {
+		for _, e := range v.Env {
+			if e.ValueFrom != nil && e.ValueFrom.SecretKeyRef != nil && !visitor(e.ValueFrom.SecretKeyRef.Name) {
+				return
+			}
+		}
+	}
+	for _, v := range pod.Spec.Containers {
+		for _, e := range v.Env {
+			if e.ValueFrom != nil && e.ValueFrom.SecretKeyRef != nil && !visitor(e.ValueFrom.SecretKeyRef.Name) {
+				return
+			}
+		}
+	}
+}
+
+func (q *restrictNodesAdmission) admitNode(nodeName string, a admission.Attributes) error {
+	// Only let nodes modify their own API resource
+	if a.GetName() == nodeName {
+		return nil
+	}
+	return kapierrors.NewForbidden(a.GetResource().GroupResource(), a.GetName(), errors.New("cannot access other nodes"))
+}
+
+func (q *restrictNodesAdmission) admitPod(nodeName string, a admission.Attributes) error {
+	if len(a.GetSubresource()) > 0 && a.GetSubresource() != "status" {
```

I think for any other sub resource you want to fail closed.